From 04f19281dfe3aa74aa48c4ea6b885b756cf10c6f Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 22:00:50 +0200
Subject: [PATCH 12/36] tvos: port profiler, vsync, ios_surface, rendering_api


diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/profiler_metrics_ios.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/profiler_metrics_ios.mm
index 0a4ba99a..16fb3db7 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/profiler_metrics_ios.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/profiler_metrics_ios.mm
@@ -58,6 +58,7 @@
 #if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG || \
     FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_PROFILE
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 std::optional<GpuUsageInfo> FindGpuUsageInfo(io_iterator_t iterator) {
   for (fml::CFRef<io_registry_entry_t> reg_entry(IOIteratorNext(iterator)); reg_entry.Get();
        reg_entry.Reset(IOIteratorNext(iterator))) {
@@ -77,18 +78,22 @@
   }
   return std::nullopt;
 }
+#endif
 
 [[maybe_unused]] std::optional<GpuUsageInfo> FindSimulatorGpuUsageInfo() {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   io_iterator_t io_iterator;
   if (IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceNameMatching("IntelAccelerator"),
                                    &io_iterator) == kIOReturnSuccess) {
     fml::CFRef<io_iterator_t> iterator(io_iterator);
     return FindGpuUsageInfo(iterator.Get());
   }
+#endif
   return std::nullopt;
 }
 
 [[maybe_unused]] std::optional<GpuUsageInfo> FindDeviceGpuUsageInfo() {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   io_iterator_t io_iterator;
   if (IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceNameMatching("sgx"),
                                    &io_iterator) == kIOReturnSuccess) {
@@ -106,6 +111,7 @@
       }
     }
   }
+#endif
   return std::nullopt;
 }
 
@@ -120,6 +126,8 @@
   return FindSimulatorGpuUsageInfo();
 #elif TARGET_OS_IOS
   return FindDeviceGpuUsageInfo();
+#elif TARGET_OS_TV
+  return std::nullopt;
 #endif  // TARGET_IPHONE_SIMULATOR
 }
 }  // namespace
diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
index e73ed315..0a4e1e91 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
@@ -96,13 +96,17 @@ - (void)setMaxRefreshRate:(double)refreshRate {
     return;
   }
   double maxFrameRate = fmax(refreshRate, 60);
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   double minFrameRate = fmax(maxFrameRate / 2, 60);
-  if (@available(iOS 15.0, *)) {
+  if (@available(iOS 15.0, tvOS 15.0, *)) {
     _displayLink.preferredFrameRateRange =
         CAFrameRateRangeMake(minFrameRate, maxFrameRate, maxFrameRate);
   } else {
     _displayLink.preferredFramesPerSecond = maxFrameRate;
   }
+#else
+  _displayLink.preferredFramesPerSecond = maxFrameRate;
+#endif
 }
 
 - (void)await {
diff --git a/engine/src/flutter/shell/platform/darwin/ios/ios_surface.mm b/engine/src/flutter/shell/platform/darwin/ios/ios_surface.mm
index b03b9b46..6c385e28 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/ios_surface.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/ios_surface.mm
@@ -21,6 +21,8 @@
   FML_DCHECK(context);
 
   if (@available(iOS METAL_IOS_VERSION_BASELINE, *)) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunguarded-availability-new"
     if ([layer isKindOfClass:[CAMetalLayer class]]) {
       switch (context->GetBackend()) {
         case IOSRenderingBackend::kSkia:
@@ -33,6 +35,7 @@
           );
       }
     }
+#pragma GCC diagnostic pop
   }
   return std::make_unique<IOSSurfaceNoop>(std::move(context));
 }
diff --git a/engine/src/flutter/shell/platform/darwin/ios/rendering_api_selection.mm b/engine/src/flutter/shell/platform/darwin/ios/rendering_api_selection.mm
index 5c77f93c..b2b7e65a 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/rendering_api_selection.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/rendering_api_selection.mm
@@ -21,10 +21,14 @@
 
 bool ShouldUseMetalRenderer() {
   bool ios_version_supports_metal = false;
+#ifdef TARGET_OS_TV
+  ios_version_supports_metal = true;  // min tvos GPU Family is same as iOS GPU family 2
+#else
   if (@available(iOS METAL_IOS_VERSION_BASELINE, *)) {
     id<MTLDevice> device = MTLCreateSystemDefaultDevice();
     ios_version_supports_metal = [device supportsFeatureSet:MTLFeatureSet_iOS_GPUFamily1_v3];
   }
+#endif
   return ios_version_supports_metal;
 }
 
@@ -63,9 +67,12 @@ Class GetCoreAnimationLayerClassForRenderingAPI(IOSRenderingAPI rendering_api) {
     case IOSRenderingAPI::kMetal:
       if (@available(iOS METAL_IOS_VERSION_BASELINE, *)) {
         if ([FlutterMetalLayer enabled]) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunguarded-availability-new"
           return [FlutterMetalLayer class];
         } else {
           return [CAMetalLayer class];
+#pragma GCC diagnostic pop
         }
       }
       FML_CHECK(false) << "Metal availability should already have been checked";
