From 057f72da7510b77f4e638928bba8bd45111152fe Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 22:17:10 +0200
Subject: [PATCH 15/36] tvos: ignore gclient state files


diff --git a/.gitignore b/.gitignore
index eb9f8e5b..be3eeabe 100644
--- a/.gitignore
+++ b/.gitignore
@@ -163,3 +163,6 @@ app.*.symbols
 .python-version
 .gclient_previous_custom_vars
 .gclient_previous_sync_commits
+.gclient
+.gclient_entries
+_bad_scm/
