From f2606c9df2298451c9f92ba3a5e9cd845fc20a83 Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 23:00:07 +0200
Subject: [PATCH 23/32] tvos: add tvos_clang_arm64/arm_sim/x64_sim toolchains


diff --git a/engine/src/build/config/BUILDCONFIG.gn b/engine/src/build/config/BUILDCONFIG.gn
index 2b0d94e1a03..69fa1efa23a 100644
--- a/engine/src/build/config/BUILDCONFIG.gn
+++ b/engine/src/build/config/BUILDCONFIG.gn
@@ -602,7 +602,17 @@ if (custom_toolchain != "") {
 } else if (is_ios) {
   import("//build/config/darwin/darwin_sdk.gni")  # For use_ios_simulator
   host_toolchain = "//build/toolchain/mac:clang_$host_cpu"
-  if (use_ios_simulator) {
+  if (is_tvos) {
+    if (use_tvos_simulator) {
+      if (target_cpu == "arm64") {
+        set_default_toolchain("//build/toolchain/mac:tvos_clang_arm_sim")
+      } else {
+        set_default_toolchain("//build/toolchain/mac:tvos_clang_x64_sim")
+      }
+    } else {
+      set_default_toolchain("//build/toolchain/mac:tvos_clang_arm64")
+    }
+  } else if (use_ios_simulator) {
     if (target_cpu == "arm64") {
       set_default_toolchain("//build/toolchain/mac:ios_clang_arm_sim")
     } else {
diff --git a/engine/src/build/toolchain/mac/BUILD.gn b/engine/src/build/toolchain/mac/BUILD.gn
index a3dfd6f2db9..f48b8f37ea7 100644
--- a/engine/src/build/toolchain/mac/BUILD.gn
+++ b/engine/src/build/toolchain/mac/BUILD.gn
@@ -428,7 +428,7 @@ template("mac_toolchain") {
   }
 }
 
-if (is_ios) {
+if (is_ios && !is_tvos) {
   # Toolchain used for iOS device targets.
   mac_toolchain("ios_clang_arm") {
     toolchain_cpu = "arm"
@@ -482,6 +482,59 @@ if (is_ios) {
   }
 }
 
+if (is_ios && is_tvos) {
+  # Toolchain used for tvOS device targets (arm64).
+  mac_toolchain("tvos_clang_arm64") {
+    toolchain_cpu = "arm64"
+    toolchain_os = "mac"
+    prefix = rebased_clang_dir
+    ar = "${prefix}/llvm-ar"
+    asm = "${prefix}/clang"
+    cc = "${cxx_prefix}${prefix}/clang"
+    cxx = "${cxx_prefix}${prefix}/clang++"
+    objc = "${objc_prefix}${prefix}/clang++"
+    ld = "${link_prefix}${prefix}/clang++"
+    is_clang = true
+
+    sysroot_rel = rebase_path(tvos_device_sdk_path, root_build_dir)
+    sysroot_flags = "-isysroot $sysroot_rel -mappletvos-version-min=$tvos_deployment_target"
+  }
+
+  # Toolchain used for tvOS simulator targets (arm64).
+  mac_toolchain("tvos_clang_arm_sim") {
+    toolchain_cpu = "arm"
+    toolchain_os = "mac"
+    prefix = rebased_clang_dir
+    ar = "${prefix}/llvm-ar"
+    asm = "${prefix}/clang"
+    cc = "${cxx_prefix}${prefix}/clang"
+    cxx = "${cxx_prefix}${prefix}/clang++"
+    objc = "${objc_prefix}${prefix}/clang++"
+    ld = "${link_prefix}${prefix}/clang++"
+    is_clang = true
+
+    sysroot_rel = rebase_path(tvos_simulator_sdk_path, root_build_dir)
+    sysroot_flags = "-isysroot $sysroot_rel -mappletvsimulator-version-min=$tvos_deployment_target"
+  }
+
+  # Toolchain used for tvOS simulator targets (x64).
+  mac_toolchain("tvos_clang_x64_sim") {
+    toolchain_cpu = "x64"
+    toolchain_os = "mac"
+    prefix = rebased_clang_dir
+    ar = "${prefix}/llvm-ar"
+    asm = "${prefix}/clang"
+    cc = "${cxx_prefix}${prefix}/clang"
+    cxx = "${cxx_prefix}${prefix}/clang++"
+    objc = "${objc_prefix}${prefix}/clang++"
+    ld = "${link_prefix}${prefix}/clang++"
+    is_clang = true
+
+    sysroot_rel = rebase_path(tvos_simulator_sdk_path, root_build_dir)
+    sysroot_flags = "-isysroot $sysroot_rel -mappletvsimulator-version-min=$tvos_deployment_target"
+  }
+}
+
 # Toolchain used for Mac X64 host targets.
 mac_toolchain("clang_x64") {
   toolchain_cpu = "x64"
