From 39a19f7032c40c0c5628a67cf0b7b4fac7dd24c9 Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Tue, 5 May 2026 22:47:20 +0200
Subject: [PATCH 35/36] tvos: fix Siri Remote directional clicks


diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
index 65d58a8c..ed5e26b5 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
@@ -1132,7 +1132,7 @@ - (void)sendGamepadTouchesWithType:(NSString*)type x:(float)x y:(float)y {
 }
 
 - (void)setupTouchesWithGamepad:(GCMicroGamepad*)gamepad {
-  gamepad.reportsAbsoluteDpadValues = false;
+  gamepad.reportsAbsoluteDpadValues = true;
   gamepad.allowsRotation = true;
   __weak FlutterViewController* weakSelf = self;
   gamepad.dpad.valueChangedHandler =
@@ -1144,17 +1144,6 @@ - (void)setupTouchesWithGamepad:(GCMicroGamepad*)gamepad {
         [strongSelf sendGamepadTouchesWithType:@"loc" x:xValue y:yValue];
         [strongSelf updateDpadWithX:xValue y:yValue];
       };
-  gamepad.buttonA.pressedChangedHandler =
-      ^(GCControllerButtonInput* _Nonnull button, float value, BOOL pressed) {
-        FlutterViewController* strongSelf = weakSelf;
-        if (!strongSelf) {
-          return;
-        }
-        [strongSelf sendSynthesizedKeyEventOfType:pressed ? kFlutterKeyEventTypeDown
-                                                   : kFlutterKeyEventTypeUp
-                                       physical:0x00070077ULL
-                                        logical:0x0010000050CULL];
-      };
   if (@available(tvOS 13.0, *)) {
     gamepad.buttonMenu.pressedChangedHandler =
         ^(GCControllerButtonInput* _Nonnull button, float value, BOOL pressed) {
