From c280f3667f1623926cdce9542bf7386ed9f4d773 Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 21:45:06 +0200
Subject: [PATCH 07/47] tvos: port FlutterTextInputPlugin


diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
index 6a49c7b8..31455b2e 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.h
@@ -30,8 +30,14 @@ typedef NS_ENUM(NSInteger, FlutterScribbleInteractionStatus) {
   // NOLINTEND(readability-identifier-naming)
 };
 
+#ifdef TARGET_OS_TV
+@interface FlutterTextInputPlugin
+    : NSObject <FlutterKeySecondaryResponder>
+#else
+
 @interface FlutterTextInputPlugin
     : NSObject <FlutterKeySecondaryResponder, UIIndirectScribbleInteractionDelegate>
+#endif
 
 @property(nonatomic, weak) UIViewController* viewController;
 @property(nonatomic, weak) id<FlutterIndirectScribbleDelegate> indirectScribbleDelegate;
@@ -137,8 +143,12 @@ API_AVAILABLE(ios(13.0)) @interface FlutterTextPlaceholder : UITextPlaceholder
 #if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
 FLUTTER_DARWIN_EXPORT
 #endif
+#ifdef TARGET_OS_TV
+@interface FlutterTextInputView : UIView <UITextInput>
+#else
 @interface FlutterTextInputView
     : UIView <UITextInput, UIScribbleInteractionDelegate, UIEditMenuInteractionDelegate>
+#endif
 
 // UITextInput
 @property(nonatomic, readonly) NSMutableString* text;
@@ -170,7 +180,9 @@ FLUTTER_DARWIN_EXPORT
 @property(nonatomic) FlutterScribbleFocusStatus scribbleFocusStatus;
 @property(nonatomic, strong) NSArray<FlutterTextSelectionRect*>* selectionRects;
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 @property(nonatomic, strong) UIEditMenuInteraction* editMenuInteraction API_AVAILABLE(ios(16.0));
+#endif
 - (void)resetScribbleInteractionStatusIfEnding;
 - (BOOL)isScribbleAvailable;
 
diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
index d454ada1..d155fea2 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm
@@ -825,7 +825,9 @@ @implementation FlutterTextInputView {
   bool _isFloatingCursorActive;
   CGPoint _floatingCursorOffset;
   bool _enableInteractiveSelection;
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   UITextInteraction* _textInteraction API_AVAILABLE(ios(13.0));
+#endif
 }
 
 @synthesize tokenizer = _tokenizer;
@@ -868,16 +870,20 @@ - (instancetype)initWithOwner:(FlutterTextInputPlugin*)textInputPlugin {
       _inlinePredictionType = UITextInlinePredictionTypeNo;
     }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
     if (@available(iOS 14.0, *)) {
       UIScribbleInteraction* interaction = [[UIScribbleInteraction alloc] initWithDelegate:self];
       [self addInteraction:interaction];
     }
+#endif
   }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (@available(iOS 16.0, *)) {
     _editMenuInteraction = [[UIEditMenuInteraction alloc] initWithDelegate:self];
     [self addInteraction:_editMenuInteraction];
   }
+#endif
 
   return self;
 }
@@ -949,6 +955,7 @@ - (void)addAdditionalBasicCommandToItems:(NSMutableArray*)items
   }
 }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (UIMenu*)editMenuInteraction:(UIEditMenuInteraction*)interaction
           menuForConfiguration:(UIEditMenuConfiguration*)configuration
               suggestedActions:(NSArray<UIMenuElement*>*)suggestedActions API_AVAILABLE(ios(16.0)) {
@@ -1054,9 +1061,12 @@ - (void)showEditMenuWithTargetRect:(CGRect)targetRect
       [UIEditMenuConfiguration configurationWithIdentifier:nil sourcePoint:CGPointZero];
   [self.editMenuInteraction presentEditMenuWithConfiguration:config];
 }
+#endif
 
 - (void)hideEditMenu API_AVAILABLE(ios(16.0)) {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [self.editMenuInteraction dismissMenu];
+#endif
 }
 
 - (void)configureWithDictionary:(NSDictionary*)configuration {
@@ -1172,6 +1182,7 @@ - (void)setTextInputClient:(int)client {
   _hasPlaceholder = NO;
 }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (UITextInteraction*)textInteraction API_AVAILABLE(ios(13.0)) {
   if (!_textInteraction) {
     _textInteraction = [UITextInteraction textInteractionForMode:UITextInteractionModeEditable];
@@ -1179,6 +1190,7 @@ - (UITextInteraction*)textInteraction API_AVAILABLE(ios(13.0)) {
   }
   return _textInteraction;
 }
+#endif
 
 - (void)setTextInputState:(NSDictionary*)state {
   // [UITextInteraction willMoveToView:] sometimes sets the textInput's inputDelegate
@@ -1186,9 +1198,11 @@ - (void)setTextInputState:(NSDictionary*)state {
   // and selection changes when that happens, add a dummy UITextInteraction to this
   // view so it sets a valid inputDelegate that we can call textWillChange et al. on.
   // See https://github.com/flutter/engine/pull/32881.
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (!self.inputDelegate && self.isFirstResponder) {
     [self addInteraction:self.textInteraction];
   }
+#endif
 
   NSString* newText = state[@"text"];
   BOOL textChanged = ![self.text isEqualToString:newText];
@@ -1226,9 +1240,11 @@ - (void)setTextInputState:(NSDictionary*)state {
     [self.inputDelegate textDidChange:self];
   }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (_textInteraction) {
     [self removeInteraction:_textInteraction];
   }
+#endif
 }
 
 // Forward touches to the viewResponder to allow tapping inside the UITextField as normal.
@@ -1295,6 +1311,7 @@ - (void)setIsVisibleToAutofill:(BOOL)isVisibleToAutofill {
 
 // Checks whether Scribble features are possibly available – meaning this is an iPad running iOS
 // 14 or higher.
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (BOOL)isScribbleAvailable {
   if (@available(iOS 14.0, *)) {
     if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
@@ -1325,6 +1342,11 @@ - (BOOL)scribbleInteractionShouldDelayFocus:(UIScribbleInteraction*)interaction
     API_AVAILABLE(ios(14.0)) {
   return NO;
 }
+#else
+- (BOOL)isScribbleAvailable {
+  return NO;
+}
+#endif
 
 #pragma mark - UIResponder Overrides
 
@@ -1348,6 +1370,7 @@ - (BOOL)resignFirstResponder {
 }
 
 - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (action == @selector(paste:)) {
     // Forbid pasting images, memojis, or other non-string content.
     return [UIPasteboard generalPasteboard].hasStrings;
@@ -1363,24 +1386,32 @@ - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
     return NO;
   }
   return [super canPerformAction:action withSender:sender];
+#endif
+  return NO;
 }
 
 #pragma mark - UIResponderStandardEditActions Overrides
 
 - (void)cut:(id)sender {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [UIPasteboard generalPasteboard].string = [self textInRange:_selectedTextRange];
   [self replaceRange:_selectedTextRange withText:@""];
+#endif
 }
 
 - (void)copy:(id)sender {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [UIPasteboard generalPasteboard].string = [self textInRange:_selectedTextRange];
+#endif
 }
 
 - (void)paste:(id)sender {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   NSString* pasteboardString = [UIPasteboard generalPasteboard].string;
   if (pasteboardString != nil) {
     [self insertText:pasteboardString];
   }
+#endif
 }
 
 - (void)delete:(id)sender {
@@ -1916,11 +1947,13 @@ - (CGRect)firstRectForRange:(UITextRange*)range {
   // API (unlike iOS 17). So we return CGRectZero to hide it (unless if scribble is enabled).
   // To support scribble's advanced gestures (e.g. insert a space with a vertical bar),
   // at least 1 character's width is required.
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (@available(iOS 17, *)) {
     // No-op
   } else if (![self isScribbleAvailable]) {
     return CGRectZero;
   }
+#endif
 
   NSUInteger first = start;
   if (end < start) {
@@ -2554,22 +2587,26 @@ - (instancetype)initWithDelegate:(id<FlutterTextInputDelegate>)textInputDelegate
     _autofillContext = [[NSMutableDictionary alloc] init];
     _inputHider = [[FlutterTextInputViewAccessibilityHider alloc] init];
     _scribbleElements = [[NSMutableDictionary alloc] init];
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
     _keyboardViewContainer = [[UIView alloc] init];
 
     [[NSNotificationCenter defaultCenter] addObserver:self
                                              selector:@selector(handleKeyboardWillShow:)
                                                  name:UIKeyboardWillShowNotification
                                                object:nil];
+#endif
   }
 
   return self;
 }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (void)handleKeyboardWillShow:(NSNotification*)notification {
   NSDictionary* keyboardInfo = [notification userInfo];
   NSValue* keyboardFrameEnd = [keyboardInfo valueForKey:UIKeyboardFrameEndUserInfoKey];
   _keyboardRect = [keyboardFrameEnd CGRectValue];
 }
+#endif
 
 - (void)dealloc {
   [self reset];
@@ -2767,12 +2804,14 @@ - (BOOL)showEditMenu:(NSDictionary*)args API_AVAILABLE(ios(16.0)) {
   if (!self.activeView.isFirstResponder) {
     return NO;
   }
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   NSDictionary<NSString*, NSNumber*>* encodedTargetRect = args[@"targetRect"];
   CGRect globalTargetRect = CGRectMake(
       [encodedTargetRect[@"x"] doubleValue], [encodedTargetRect[@"y"] doubleValue],
       [encodedTargetRect[@"width"] doubleValue], [encodedTargetRect[@"height"] doubleValue]);
   CGRect localTargetRect = [self.hostView convertRect:globalTargetRect toView:self.activeView];
   [self.activeView showEditMenuWithTargetRect:localTargetRect items:args[@"items"]];
+#endif
   return YES;
 }
 
@@ -2783,8 +2822,11 @@ - (void)hideEditMenu {
 - (void)setEditableSizeAndTransform:(NSDictionary*)dictionary {
   NSArray* transform = dictionary[@"transform"];
   [_activeView setEditableTransform:transform];
+
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   const int leftIndex = 12;
   const int topIndex = 13;
+
   if ([_activeView isScribbleAvailable]) {
     // This is necessary to set up where the scribble interactable element will be.
     _inputHider.frame =
@@ -2808,6 +2850,7 @@ - (void)setEditableSizeAndTransform:(NSDictionary*)dictionary {
           CGRectMake([transform[leftIndex] intValue], [transform[topIndex] intValue], 0, 0);
     }
   }
+#endif
 }
 
 - (void)updateMarkedRect:(NSDictionary*)dictionary {
@@ -2842,12 +2885,14 @@ - (void)setSelectionRects:(NSArray*)encodedRects {
 }
 
 - (void)startLiveTextInput {
-  if (@available(iOS 15.0, *)) {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
+  if (@available(iOS 15.0, tvOS 15.0, *)) {
     if (_activeView == nil || !_activeView.isFirstResponder) {
       return;
     }
     [_activeView captureTextFromCamera:nil];
   }
+#endif
 }
 
 - (void)showTextInput {
@@ -3204,6 +3249,7 @@ - (void)updateConfig:(NSDictionary*)dictionary {
 
 #pragma mark UIIndirectScribbleInteractionDelegate
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (BOOL)indirectScribbleInteraction:(UIIndirectScribbleInteraction*)interaction
                    isElementFocused:(UIScribbleElementIdentifier)elementIdentifier
     API_AVAILABLE(ios(14.0)) {
@@ -3279,11 +3325,13 @@ - (void)indirectScribbleInteraction:(UIIndirectScribbleInteraction*)interaction
                         completion(elements);
                       }];
 }
+#endif
 
 #pragma mark - Methods related to Scribble support
 
 - (void)setUpIndirectScribbleInteraction:(id<FlutterViewResponder>)viewResponder {
   if (_viewResponder != viewResponder) {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
     if (@available(iOS 14.0, *)) {
       UIView* parentView = viewResponder.view;
       if (parentView != nil) {
@@ -3292,6 +3340,7 @@ - (void)setUpIndirectScribbleInteraction:(id<FlutterViewResponder>)viewResponder
         [parentView addInteraction:scribbleInteraction];
       }
     }
+#endif
   }
   _viewResponder = viewResponder;
 }
