From a2c6452a423263ff4a5d77ecf60b8b73024a8750 Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 21:53:45 +0200
Subject: [PATCH 09/47] tvos: port FlutterViewController.mm


diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
index 831568de..8de0d5cd 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
@@ -9,6 +9,11 @@
 #import <os/log.h>
 #include <memory>
 
+#ifdef TARGET_OS_TV
+#include <GameController/GameController.h>
+#include <MediaPlayer/MediaPlayer.h>
+#endif
+
 #include "flutter/common/constants.h"
 #include "flutter/fml/memory/weak_ptr.h"
 #include "flutter/fml/message_loop.h"
@@ -37,6 +42,9 @@
 #import "flutter/shell/platform/darwin/ios/platform_view_ios.h"
 #import "flutter/shell/platform/embedder/embedder.h"
 #import "flutter/third_party/spring_animation/spring_animation.h"
+#if defined(TARGET_OS_TV) && TARGET_OS_TV
+#import "flutter/shell/platform/darwin/ios/framework/Source/FlutterAccessibilitySelectionView.h"
+#endif
 
 FLUTTER_ASSERT_ARC
 
@@ -64,7 +72,13 @@
 // This is left a FlutterBinaryMessenger privately for now to give people a chance to notice the
 // change. Unfortunately unless you have Werror turned on, incompatible pointers as arguments are
 // just a warning.
+#ifdef TARGET_OS_TV
+@interface FlutterViewController () <FlutterBinaryMessenger, UIScrollViewDelegate, UIGestureRecognizerDelegate>
+@property(nonatomic, strong) FlutterBasicMessageChannel* keyEventChannel;
+@property(nonatomic, strong) FlutterBasicMessageChannel* gamepadTouchEventChannel;
+#else
 @interface FlutterViewController () <FlutterBinaryMessenger, UIScrollViewDelegate>
+#endif
 // TODO(dkwingsmt): Make the view ID property public once the iOS shell
 // supports multiple views.
 // https://github.com/flutter/flutter/issues/138168
@@ -85,6 +99,10 @@ @interface FlutterViewController () <FlutterBinaryMessenger, UIScrollViewDelegat
 @property(nonatomic, assign) BOOL isHomeIndicatorHidden;
 @property(nonatomic, assign) BOOL isPresentingViewControllerAnimating;
 
+#ifdef TARGET_OS_TV
+@property(nonatomic, strong) FlutterAccessibilitySelectionView *selectionView;
+#endif
+
 // Internal state backing override of UIView.prefersStatusBarHidden.
 @property(nonatomic, assign) BOOL flutterPrefersStatusBarHidden;
 
@@ -138,6 +156,7 @@ @interface FlutterViewController () <FlutterBinaryMessenger, UIScrollViewDelegat
  * Mouse and trackpad gesture recognizers
  */
 // Mouse and trackpad hover
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 @property(nonatomic, strong)
     UIHoverGestureRecognizer* hoverGestureRecognizer API_AVAILABLE(ios(13.4));
 // Mouse wheel scrolling
@@ -152,6 +171,7 @@ @interface FlutterViewController () <FlutterBinaryMessenger, UIScrollViewDelegat
 // Trackpad rotating
 @property(nonatomic, strong)
     UIRotationGestureRecognizer* rotationGestureRecognizer API_AVAILABLE(ios(13.4));
+#endif
 
 /// Creates and registers plugins used by this view controller.
 - (void)addInternalPlugins;
@@ -357,8 +377,10 @@ - (void)performCommonViewControllerInitialization {
   }
 
   _initialized = YES;
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   _orientationPreferences = UIInterfaceOrientationMaskAll;
   _statusBarStyle = UIStatusBarStyleDefault;
+#endif
 
   _accessibilityFeatures = [[FlutterAccessibilityFeatures alloc] init];
 
@@ -369,6 +391,7 @@ - (void)performCommonViewControllerInitialization {
 
 - (void)setUpNotificationCenterObservers {
   NSNotificationCenter* center = [NSNotificationCenter defaultCenter];
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [center addObserver:self
              selector:@selector(onOrientationPreferencesUpdated:)
                  name:@(flutter::kOrientationUpdateNotificationName)
@@ -378,6 +401,7 @@ - (void)setUpNotificationCenterObservers {
              selector:@selector(onPreferredStatusBarStyleUpdated:)
                  name:@(flutter::kOverlayStyleUpdateNotificationName)
                object:nil];
+#endif
 
   if (FlutterSharedApplication.isAvailable) {
     [self setUpApplicationLifecycleNotifications:center];
@@ -385,6 +409,7 @@ - (void)setUpNotificationCenterObservers {
     [self setUpSceneLifecycleNotifications:center];
   }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [center addObserver:self
              selector:@selector(keyboardWillChangeFrame:)
                  name:UIKeyboardWillChangeFrameNotification
@@ -399,6 +424,7 @@ - (void)setUpNotificationCenterObservers {
              selector:@selector(keyboardWillBeHidden:)
                  name:UIKeyboardWillHideNotification
                object:nil];
+#endif
 
   for (NSString* notification in [self.accessibilityFeatures observedNotificationNames]) {
     [center addObserver:self
@@ -412,6 +438,7 @@ - (void)setUpNotificationCenterObservers {
                  name:UIContentSizeCategoryDidChangeNotification
                object:nil];
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [center addObserver:self
              selector:@selector(onHideHomeIndicatorNotification:)
                  name:FlutterViewControllerHideHomeIndicator
@@ -421,6 +448,7 @@ - (void)setUpNotificationCenterObservers {
              selector:@selector(onShowHomeIndicatorNotification:)
                  name:FlutterViewControllerShowHomeIndicator
                object:nil];
+#endif
 }
 
 - (void)setUpSceneLifecycleNotifications:(NSNotificationCenter*)center API_AVAILABLE(ios(13.0)) {
@@ -471,10 +499,12 @@ - (void)setUpApplicationLifecycleNotifications:(NSNotificationCenter*)center {
                  name:UIApplicationDidEnterBackgroundNotification
                object:nil];
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [center addObserver:self
              selector:@selector(applicationWillEnterForeground:)
                  name:UIApplicationWillEnterForegroundNotification
                object:nil];
+#endif
 }
 
 - (void)setInitialRoute:(NSString*)route {
@@ -499,7 +529,11 @@ - (void)pushRoute:(NSString*)route {
   auto placeholder = [[UIView alloc] init];
 
   placeholder.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   placeholder.backgroundColor = UIColor.systemBackgroundColor;
+#else
+  placeholder.backgroundColor = UIColor.whiteColor;
+#endif
   placeholder.autoresizesSubviews = YES;
 
   // Only add the label when we know we have failed to enable tracing (and it was necessary).
@@ -523,7 +557,9 @@ - (void)pushRoute:(NSString*)route {
 
 - (void)loadView {
   self.view = GetViewOrPlaceholder(self.flutterView);
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   self.view.multipleTouchEnabled = YES;
+#endif
   self.view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
 
   [self installSplashScreenViewIfNecessary];
@@ -787,11 +823,213 @@ - (void)surfaceUpdated:(BOOL)appeared {
   }
 }
 
+#pragma mark - remote tap handling
+#ifdef TARGET_OS_TV
+- (void)sendTap:(int)keyCode withType:(NSString*)keyMapType ofType:(NSString*)type {
+  NSMutableDictionary* keyMessage = [@{
+    @"keymap" : keyMapType,
+    @"type" : type,
+    @"keyCode" : @(keyCode),
+    @"modifiers" : @(0),
+  } mutableCopy];
+  [self.keyEventChannel sendMessage:keyMessage];
+}
+- (void)handleTap:(UITapGestureRecognizer*)recognizer
+         withType:(NSString*)keyMapType
+          keyType:(int)key {
+  if (recognizer.state == UIGestureRecognizerStateBegan) {
+    [self sendTap:key withType:keyMapType ofType:@"keydown"];
+  } else if (recognizer.state == UIGestureRecognizerStateEnded) {
+    [self sendTap:key withType:keyMapType ofType:@"keyup"];
+  }
+}
+- (void)handleDownTap:(UITapGestureRecognizer*)recognizer {
+  [self handleTap:recognizer withType:@"macos" keyType:0x7D];
+}
+- (void)handleUpTap:(UITapGestureRecognizer*)recognizer {
+  [self handleTap:recognizer withType:@"macos" keyType:0x7E];
+}
+- (void)handleLeftTap:(UITapGestureRecognizer*)recognizer {
+  [self handleTap:recognizer withType:@"macos" keyType:0x7B];
+}
+- (void)handleRightTap:(UITapGestureRecognizer*)recognizer {
+  [self handleTap:recognizer withType:@"macos" keyType:0x7C];
+}
+- (void)handleCenterTap:(UITapGestureRecognizer*)recognizer {
+  // only send once
+  if (recognizer.state == UIGestureRecognizerStateBegan) {
+    [self sendGamepadTouchesWithType:@"click_s" x:0 y:0];
+  } else if (recognizer.state == UIGestureRecognizerStateEnded) {
+    [self sendGamepadTouchesWithType:@"click_e" x:0 y:0];
+  }
+}
+- (void)handlePlayPauseTap:(UITapGestureRecognizer*)recognizer {
+  // No such key on macOS so we fake android
+  [self handleTap:recognizer withType:@"android" keyType:0x55];
+}
+- (void)handleMenuTap:(UITapGestureRecognizer*)recognizer {
+  // pop route should be triggered on UI, to have correct behavior
+  [self handleTap:recognizer withType:@"android" keyType:0x04];
+}
+- (void)handlePageUpTap:(UITapGestureRecognizer*)recognizer API_AVAILABLE(tvos(14.3)) {
+  [self handleTap:recognizer withType:@"android" keyType:0xA6];
+}
+- (void)handlePageDownTap:(UITapGestureRecognizer*)recognizer API_AVAILABLE(tvos(14.3)) {
+  [self handleTap:recognizer withType:@"android" keyType:0xA7];
+}
+- (void)createRecognizerFor:(UIPressType)pressType action:(nullable SEL)action {
+  UILongPressGestureRecognizer* tapGestureRecognizer =
+      [[UILongPressGestureRecognizer alloc] initWithTarget:self action:action];
+  tapGestureRecognizer.allowedPressTypes = @[ @(pressType) ];
+  tapGestureRecognizer.minimumPressDuration = 0.0;
+  tapGestureRecognizer.delegate = self;
+  [self.view addGestureRecognizer:tapGestureRecognizer];
+}
+
+- (void)sendGamepadTouchesWithType:(NSString*)type x:(float)x y:(float)y {
+  NSMutableDictionary* dictionary = [@{
+    @"type" : type,
+    @"x" : @(x),
+    @"y" : @(y),
+  } mutableCopy];
+  [self.gamepadTouchEventChannel sendMessage:dictionary];
+}
+
+- (void)setupTouchesWithGamepad:(GCMicroGamepad*)gamepad {
+  gamepad.reportsAbsoluteDpadValues = true;
+  gamepad.dpad.valueChangedHandler =
+      ^(GCControllerDirectionPad* _Nonnull dpad, float xValue, float yValue) {
+        [self sendGamepadTouchesWithType:@"loc" x:xValue y:yValue];
+      };
+}
+
+- (void)setupControllers {  // TODO: dangerous, multiple registrations can happen!
+  bool atLeastOne = false;
+  for (GCController* controller in [GCController controllers]) {
+    GCMicroGamepad* gamepad = controller.microGamepad;
+    if (gamepad != nil) {
+      atLeastOne = true;
+      [self setupTouchesWithGamepad:gamepad];
+    }
+  }
+  if (!atLeastOne) {
+    NSLog(@"No controllers found");
+  }
+}
+
+- (void)setupWalnutController {
+  MPRemoteCommandCenter* commandCenter = [MPRemoteCommandCenter sharedCommandCenter];
+
+  [commandCenter.seekForwardCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        if (((MPSeekCommandEvent*)event).type == 0) {
+          [self sendTap:0x5A withType:@"android" ofType:@"keydown"];
+        } else if (((MPSeekCommandEvent*)event).type == 1) {
+          [self sendTap:0x5A withType:@"android" ofType:@"keyup"];
+        }
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.seekBackwardCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        if (((MPSeekCommandEvent*)event).type == 0) {
+          [self sendTap:0x59 withType:@"android" ofType:@"keydown"];
+        } else if (((MPSeekCommandEvent*)event).type == 1) {
+          [self sendTap:0x59 withType:@"android" ofType:@"keyup"];
+        }
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.playCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        [self sendTap:0x7E withType:@"android" ofType:@"keydown"];
+        [self sendTap:0x7E withType:@"android" ofType:@"keyup"];
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.pauseCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        [self sendTap:0x7F withType:@"android" ofType:@"keydown"];
+        [self sendTap:0x7F withType:@"android" ofType:@"keyup"];
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.stopCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        [self sendTap:0x56 withType:@"android" ofType:@"keydown"];
+        [self sendTap:0x56 withType:@"android" ofType:@"keyup"];
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.togglePlayPauseCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        [self sendTap:0x55 withType:@"android" ofType:@"keydown"];
+        [self sendTap:0x55 withType:@"android" ofType:@"keyup"];
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+
+  [commandCenter.changePlaybackRateCommand
+      addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent* event) {
+        if (((MPChangePlaybackRateCommandEvent*)event).playbackRate > 0) {
+          [self sendTap:0x5A withType:@"android" ofType:@"keydown"];
+          [self sendTap:0x5A withType:@"android" ofType:@"keyup"];
+        } else if (((MPChangePlaybackRateCommandEvent*)event).playbackRate < 0) {
+          [self sendTap:0x59 withType:@"android" ofType:@"keydown"];
+          [self sendTap:0x59 withType:@"android" ofType:@"keyup"];
+        }
+        return MPRemoteCommandHandlerStatusSuccess;
+      }];
+}
+
+- (void)controllerConnected:(NSNotification*)notification {
+  [self setupControllers];
+}
+#endif  // TARGET_OS_TV
+
 #pragma mark - UIViewController lifecycle notifications
 
 - (void)viewDidLoad {
   TRACE_EVENT0("flutter", "viewDidLoad");
 
+#ifdef TARGET_OS_TV
+  [self createRecognizerFor:UIPressTypeUpArrow action:@selector(handleUpTap:)];
+  [self createRecognizerFor:UIPressTypeDownArrow action:@selector(handleDownTap:)];
+  [self createRecognizerFor:UIPressTypeLeftArrow action:@selector(handleLeftTap:)];
+  [self createRecognizerFor:UIPressTypeRightArrow action:@selector(handleRightTap:)];
+  [self createRecognizerFor:UIPressTypeSelect action:@selector(handleCenterTap:)];
+  [self createRecognizerFor:UIPressTypePlayPause action:@selector(handlePlayPauseTap:)];
+  [self createRecognizerFor:UIPressTypeMenu action:@selector(handleMenuTap:)];
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+
+  if ([self respondsToSelector:@selector(handlePageUpTap:)]) {
+    [self createRecognizerFor:UIPressTypePageUp action:@selector(handlePageUpTap:)];
+  }
+
+  if ([self respondsToSelector:@selector(handlePageDownTap:)]) {
+    [self createRecognizerFor:UIPressTypePageDown action:@selector(handlePageDownTap:)];
+  }
+
+#pragma clang diagnostic pop
+
+  [self setupControllers];
+  [self setupWalnutController];
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(controllerConnected:)
+                                               name:GCControllerDidConnectNotification
+                                             object:nil];
+
+  self.keyEventChannel = [FlutterBasicMessageChannel
+      messageChannelWithName:@"flutter/keyevent"
+             binaryMessenger:[self binaryMessenger]
+                       codec:[FlutterJSONMessageCodec sharedInstance]];
+  self.gamepadTouchEventChannel = [FlutterBasicMessageChannel
+      messageChannelWithName:@"flutter/gamepadtouchevent"
+             binaryMessenger:[self binaryMessenger]
+                       codec:[FlutterJSONMessageCodec sharedInstance]];
+#endif  // TARGET_OS_TV
+
   if (self.engine && self.engineNeedsLaunch) {
     [self.engine launchEngine:nil libraryURI:nil entrypointArgs:nil];
     [self.engine setViewController:self];
@@ -806,6 +1044,7 @@ - (void)viewDidLoad {
   // Create a vsync client to correct delivery frame rate of touch events if needed.
   [self createTouchRateCorrectionVSyncClientIfNeeded];
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (@available(iOS 13.4, *)) {
     _hoverGestureRecognizer =
         [[UIHoverGestureRecognizer alloc] initWithTarget:self action:@selector(hoverEvent:)];
@@ -839,8 +1078,14 @@ - (void)viewDidLoad {
     _rotationGestureRecognizer.delegate = self;
     [self.flutterView addGestureRecognizer:_rotationGestureRecognizer];
   }
+#endif
 
   [super viewDidLoad];
+
+#if defined(TARGET_OS_TV) && TARGET_OS_TV
+  self.selectionView = [[FlutterAccessibilitySelectionView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
+  [self.view addSubview:self.selectionView];
+#endif
 }
 
 - (void)addInternalPlugins {
@@ -910,8 +1155,10 @@ - (void)viewWillDisappear:(BOOL)animated {
 - (void)viewDidDisappear:(BOOL)animated {
   TRACE_EVENT0("flutter", "viewDidDisappear");
   if (self.engine.viewController == self) {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
     [self invalidateKeyboardAnimationVSyncClient];
     [self ensureViewportMetricsIsCorrect];
+#endif
     [self surfaceUpdated:NO];
     [self.engine.lifecycleChannel sendMessage:@"AppLifecycleState.paused"];
     [self flushOngoingTouches];
@@ -1000,17 +1247,21 @@ - (void)dealloc {
   [self removeInternalPlugins];
   [self deregisterNotifications];
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [self invalidateKeyboardAnimationVSyncClient];
+#endif
   [self invalidateTouchRateCorrectionVSyncClient];
 
   // TODO(cbracken): https://github.com/flutter/flutter/issues/156222
   // Ensure all delegates are weak and remove this.
   _scrollView.delegate = nil;
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   _hoverGestureRecognizer.delegate = nil;
   _discreteScrollingPanGestureRecognizer.delegate = nil;
   _continuousScrollingPanGestureRecognizer.delegate = nil;
   _pinchGestureRecognizer.delegate = nil;
   _rotationGestureRecognizer.delegate = nil;
+#endif
 }
 
 #pragma mark - Application lifecycle notifications
@@ -1293,6 +1544,7 @@ - (void)dispatchTouches:(NSSet*)touches
     // Sweep direction is the same. Phase of M_PI_2.
     pointer_data.orientation = [touch azimuthAngleInView:nil] - M_PI_2;
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
     if (@available(iOS 13.4, *)) {
       if (event != nullptr) {
         pointer_data.buttons = (((event.buttonMask & UIEventButtonMaskPrimary) > 0)
@@ -1303,6 +1555,7 @@ - (void)dispatchTouches:(NSSet*)touches
                                     : 0);
       }
     }
+#endif
 
     packet->SetPointerData(pointer_index++, pointer_data);
 
@@ -1317,19 +1570,47 @@ - (void)dispatchTouches:(NSSet*)touches
 }
 
 - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
+#ifdef TARGET_OS_TV
+  for (UITouch* touch in touches) {
+    CGPoint location = [touch locationInView:self.view];
+    [self sendGamepadTouchesWithType:@"started" x:location.x y:location.y];
+  }
+#else
   [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
+#endif
 }
 
 - (void)touchesMoved:(NSSet*)touches withEvent:(UIEvent*)event {
+#ifdef TARGET_OS_TV
+  for (UITouch* touch in touches) {
+    CGPoint location = [touch locationInView:self.view];
+    [self sendGamepadTouchesWithType:@"move" x:location.x y:location.y];
+  }
+#else
   [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
+#endif
 }
 
 - (void)touchesEnded:(NSSet*)touches withEvent:(UIEvent*)event {
+#ifdef TARGET_OS_TV
+  for (UITouch* touch in touches) {
+    CGPoint location = [touch locationInView:self.view];
+    [self sendGamepadTouchesWithType:@"ended" x:location.x y:location.y];
+  }
+#else
   [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
+#endif
 }
 
 - (void)touchesCancelled:(NSSet*)touches withEvent:(UIEvent*)event {
+#ifdef TARGET_OS_TV
+  for (UITouch* touch in touches) {
+    CGPoint location = [touch locationInView:self.view];
+    [self sendGamepadTouchesWithType:@"cancelled" x:location.x y:location.y];
+  }
+#else
   [self dispatchTouches:touches pointerDataChangeOverride:nullptr event:event];
+#endif
 }
 
 - (void)forceTouchesCancelled:(NSSet*)touches {
@@ -1441,6 +1722,11 @@ - (void)viewDidLayoutSubviews {
                        }
                      }];
   }
+#if defined(TARGET_OS_TV) && TARGET_OS_TV
+  // the selection view must remain the top most view, otherwise the accessibility selection
+  // rectangle drawn by the system is not visible!
+  [self.view bringSubviewToFront:self.selectionView];
+#endif
 }
 
 - (BOOL)isAutoResizable {
@@ -1567,6 +1853,8 @@ - (void)setViewportMetricsPaddings {
 
 #pragma mark - Keyboard events
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
+
 - (void)keyboardWillShowNotification:(NSNotification*)notification {
   // Immediately prior to a docked keyboard being shown or when a keyboard goes from
   // undocked/floating to docked, this notification is triggered. This notification also happens
@@ -1958,6 +2246,7 @@ - (void)ensureViewportMetricsIsCorrect {
     [self updateViewportMetricsIfNeeded];
   }
 }
+#endif  // !TARGET_OS_TV
 
 - (void)handlePressEvent:(FlutterUIPressProxy*)press
               nextAction:(void (^)())next API_AVAILABLE(ios(13.4)) {
@@ -2003,9 +2292,10 @@ - (void)superPressesCancelled:(NSSet<UIPress*>*)presses withEvent:(UIPressesEven
 // both places to capture keys both inside and outside of a text field, but have
 // slightly different implementations.
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (void)pressesBegan:(NSSet<UIPress*>*)presses
-           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0)) {
-  if (@available(iOS 13.4, *)) {
+           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  if (@available(iOS 13.4, tvOS 13.4, *)) {
     __weak FlutterViewController* weakSelf = self;
     for (UIPress* press in presses) {
       [self handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press event:event]
@@ -2019,8 +2309,8 @@ - (void)pressesBegan:(NSSet<UIPress*>*)presses
 }
 
 - (void)pressesChanged:(NSSet<UIPress*>*)presses
-             withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0)) {
-  if (@available(iOS 13.4, *)) {
+             withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  if (@available(iOS 13.4, tvOS 13.4, *)) {
     __weak FlutterViewController* weakSelf = self;
     for (UIPress* press in presses) {
       [self handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press event:event]
@@ -2034,8 +2324,8 @@ - (void)pressesChanged:(NSSet<UIPress*>*)presses
 }
 
 - (void)pressesEnded:(NSSet<UIPress*>*)presses
-           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0)) {
-  if (@available(iOS 13.4, *)) {
+           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  if (@available(iOS 13.4, tvOS 13.4, *)) {
     __weak FlutterViewController* weakSelf = self;
     for (UIPress* press in presses) {
       [self handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press event:event]
@@ -2049,8 +2339,8 @@ - (void)pressesEnded:(NSSet<UIPress*>*)presses
 }
 
 - (void)pressesCancelled:(NSSet<UIPress*>*)presses
-               withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0)) {
-  if (@available(iOS 13.4, *)) {
+               withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  if (@available(iOS 13.4, tvOS 13.4, *)) {
     __weak FlutterViewController* weakSelf = self;
     for (UIPress* press in presses) {
       [self handlePressEvent:[[FlutterUIPressProxy alloc] initWithPress:press event:event]
@@ -2062,9 +2352,32 @@ - (void)pressesCancelled:(NSSet<UIPress*>*)presses
     [super pressesCancelled:presses withEvent:event];
   }
 }
+#else  // TARGET_OS_TV
+
+- (void)pressesBegan:(NSSet<UIPress*>*)presses
+           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  [super pressesBegan:presses withEvent:event];
+}
+
+- (void)pressesChanged:(NSSet<UIPress*>*)presses
+             withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  [super pressesChanged:presses withEvent:event];
+}
+
+- (void)pressesEnded:(NSSet<UIPress*>*)presses
+           withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  [super pressesEnded:presses withEvent:event];
+}
+
+- (void)pressesCancelled:(NSSet<UIPress*>*)presses
+               withEvent:(UIPressesEvent*)event API_AVAILABLE(ios(9.0), tvos(10.0)) {
+  [super pressesCancelled:presses withEvent:event];
+}
+#endif  // TARGET_OS_TV
 
 #pragma mark - Orientation updates
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (void)onOrientationPreferencesUpdated:(NSNotification*)notification {
   // Notifications may not be on the iOS UI thread
   __weak FlutterViewController* weakSelf = self;
@@ -2098,7 +2411,7 @@ - (void)performOrientationUpdate:(UIInterfaceOrientationMask)new_preferences {
   if (new_preferences != self.orientationPreferences) {
     self.orientationPreferences = new_preferences;
 
-    if (@available(iOS 16.0, *)) {
+    if (@available(iOS 16.0, tvOS 16.0, *)) {
       UIApplication* flutterApplication = FlutterSharedApplication.application;
       NSSet<UIScene*>* scenes = [NSSet set];
       if (flutterApplication) {
@@ -2163,6 +2476,7 @@ - (void)setIsHomeIndicatorHidden:(BOOL)hideHomeIndicator {
 - (BOOL)prefersHomeIndicatorAutoHidden {
   return self.isHomeIndicatorHidden;
 }
+#endif  // !TARGET_OS_TV
 
 - (BOOL)shouldAutorotate {
   return YES;
@@ -2285,6 +2599,22 @@ - (CGFloat)textScaleFactor {
   }
 }
 
+#if defined(TARGET_OS_TV) && TARGET_OS_TV
+- (UIView*)getAccessibilitySelectionVliew {
+  return (UIView*)self.selectionView;
+}
+
+- (NSArray<id<UIFocusEnvironment>>*)preferredFocusEnvironments {
+  // focus must remain on selection view, otherwise the selectangle from the system will not be
+  // visible
+  if (self.selectionView.hidden == NO) {
+    return @[ self.selectionView ];
+  } else {
+    return @[];
+  }
+}
+#endif
+
 - (BOOL)supportsShowingSystemContextMenu {
   if (@available(iOS 16.0, *)) {
     return YES;
@@ -2310,6 +2640,9 @@ - (NSString*)brightnessMode {
 // understood by the Flutter framework. See the settings system channel for more
 // information.
 - (NSString*)contrastMode {
+#ifdef TARGET_OS_TV
+  return @"normal";
+#else
   UIAccessibilityContrast contrast = self.traitCollection.accessibilityContrast;
 
   if (contrast == UIAccessibilityContrastHigh) {
@@ -2317,10 +2650,12 @@ - (NSString*)contrastMode {
   } else {
     return @"normal";
   }
+#endif
 }
 
 #pragma mark - Status bar style
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (UIStatusBarStyle)preferredStatusBarStyle {
   return self.statusBarStyle;
 }
@@ -2358,6 +2693,7 @@ - (void)setPrefersStatusBarHidden:(BOOL)hidden {
 - (BOOL)prefersStatusBarHidden {
   return self.flutterPrefersStatusBarHidden;
 }
+#endif  // !TARGET_OS_TV
 
 #pragma mark - Platform views
 
@@ -2490,6 +2826,7 @@ - (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
 
 - (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
        shouldReceiveEvent:(UIEvent*)event API_AVAILABLE(ios(13.4)) {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   if (gestureRecognizer == _continuousScrollingPanGestureRecognizer &&
       event.type == UIEventTypeScroll) {
     // Events with type UIEventTypeScroll are only received when running on macOS under emulation.
@@ -2508,10 +2845,12 @@ - (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
       self.scrollInertiaEventAppKitDeadline = 0;
     }
   }
+#endif
   // This method is also called for UITouches, should return YES to process all touches.
   return YES;
 }
 
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
 - (void)hoverEvent:(UIHoverGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
   CGPoint oldLocation = _mouseState.location;
 
@@ -2679,6 +3018,7 @@ - (void)pinchEvent:(UIPinchGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)
   packet->SetPointerData(/*i=*/0, pointer_data);
   [self.engine dispatchPointerDataPacket:std::move(packet)];
 }
+#endif  // !TARGET_OS_TV
 
 #pragma mark - State Restoration
 
