From bb337c0af8ff05f841a825bdf9c29d6534533b42 Mon Sep 17 00:00:00 2001
From: edde746 <86283021+edde746@users.noreply.github.com>
Date: Thu, 23 Apr 2026 23:15:30 +0200
Subject: [PATCH 31/40] tvos: guard share/search/lookup calls in FlutterEngine


diff --git a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
index 6b4969ab..c3143eb3 100644
--- a/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
+++ b/engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
@@ -1131,17 +1131,23 @@ - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
 
 - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
            shareSelectedText:(NSString*)selectedText {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [self.platformPlugin showShareViewController:selectedText];
+#endif
 }
 
 - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
     searchWebWithSelectedText:(NSString*)selectedText {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [self.platformPlugin searchWeb:selectedText];
+#endif
 }
 
 - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
           lookUpSelectedText:(NSString*)selectedText {
+#if !(defined(TARGET_OS_TV) && TARGET_OS_TV)
   [self.platformPlugin showLookUpViewController:selectedText];
+#endif
 }
 
 - (void)flutterTextInputView:(FlutterTextInputView*)textInputView
