From d81ea698e9d27b5f2466840e703388479cbefa53 Mon Sep 17 00:00:00 2001
From: tvOS port <tvos@local>
Date: Mon, 18 May 2026 20:40:49 +0200
Subject: [PATCH 39/51] tvos: reflect metal-tvos impeller shaders


diff --git a/engine/src/flutter/impeller/tools/compiler.gni b/engine/src/flutter/impeller/tools/compiler.gni
index 7435a730..fc5cdabf 100644
--- a/engine/src/flutter/impeller/tools/compiler.gni
+++ b/engine/src/flutter/impeller/tools/compiler.gni
@@ -33,6 +33,7 @@ template("_impellerc") {
 # Required: shader_target_flags     The target flag(s) to append. Valid options:
 #                                       --sksl
 #                                       --metal-ios
+#                                       --metal-tvos
 #                                       --metal-desktop
 #                                       --opengl-es
 #                                       --opengl-desktop
@@ -194,6 +195,7 @@ template("impellerc") {
       if (filter_include(shader_target_flags,
                          [
                            "--metal-ios",
+                           "--metal-tvos",
                            "--metal-desktop",
                            "--opengl-es",
                            "--opengl-desktop",
